src/task/Poller.ts

```diff
@@ -32,87 +33,62 @@ export class Poller {
     return this.polling;
   }
 
+  get tasksInProcess() {
+    return this._tasksInProcess;
+  }
+
   /**
    * Starts polling for work
    */
   startPolling = () => {
     if (this.polling) {
       throw new Error("Runner is already started");
     }
-
-    return this.poll();
+    this._tasksInProcess = 0;
+    this.polling = true;
+    this.poll();
   };
 
   /**
    * Stops Polling for work
    */
   stopPolling = async () => {
-    await Promise.all(this.concurrentCalls.map((call) => call.stop()));
     this.polling = false;
+    clearTimeout(this.timeoutHandler!);
   };
 
-  /**
-   * adds or shuts down concurrent calls based on the concurrency setting
-   * @param concurrency
-   */
-  private updateConcurrency(concurrency: number) {
-    if (concurrency > 0 && concurrency !== this.options.concurrency) {
-      if (concurrency < this.options.concurrency) {
-        const result = this.concurrentCalls.splice(
-          0,
-          this.options.concurrency - concurrency
-        );
-        result.forEach((call) => {
-          call.stop();
-          this.logger.debug("stopping some spawned calls");
-        });
-      } else {
-        for (let i = 0; i < concurrency - this.options.concurrency; i++) {
-          this.concurrentCalls.push(this.singlePoll());
-          this.logger.debug("spawning additional poll calls");
-        }
-      }
-      this.options.concurrency = concurrency;
-    }
-  }
+  private performWork = async (work: T) => {
+    await this.performWorkFunction(work);
+    this._tasksInProcess--;
+  };
 
   updateOptions(options: Partial<PollerOptions>) {
     const newOptions = { ...this.options, ...options };
-    this.updateConcurrency(newOptions.concurrency);
     this.options = newOptions;
   }
 
   private poll = async () => {
-    if (!this.polling) {
-      this.polling = true;
-      for (let i = 0; i < this.options.concurrency; i++) {
-        this.concurrentCalls.push(this.singlePoll());
-      }
-    }
-  };
+    while (this.isPolling) {
+      // Concurrency could have been updated. Accounting for that
+      const count =
+        this.options.concurrency > this._tasksInProcess
+          ? this.options.concurrency - this._tasksInProcess
+          : this._tasksInProcess;
```

Comment:
const count = Math.max(0, this.options.concurrency - this._tasksInProcess);

// Math.max(0, -10) = 0
// Math.max(0, 1) = 1